src/adagrams.js

```diff
@@ -1,15 +1,183 @@
+  N: 6,
+  O: 8,
+  P: 2,
+  Q: 1,
+  R: 6,
+  S: 4,
+  T: 6,
+  U: 4,
+  V: 2,
+  W: 2,
+  X: 1,
+  Y: 2,
+  Z: 1,
+};
+
+const scoreChart = {
+  A: 1,
+  B: 3,
+  C: 3,
+  D: 2,
+  E: 1,
+  F: 4,
+  G: 2,
+  H: 4,
+  I: 1,
+  J: 8,
+  K: 5,
+  L: 1,
+  M: 3,
+  N: 1,
+  O: 1,
+  P: 3,
+  Q: 10,
+  R: 1,
+  S: 1,
+  T: 1,
+  U: 1,
+  V: 4,
+  W: 4,
+  X: 8,
+  Y: 4,
+  Z: 10,
+};
+
+// converts letterPool into one array of letters according to their quantity
+const letterSoup = (letterPool) => {
+  let letterArray = [];
+
+  for (const letter in letterPool) {
+    for (let runner = 0; runner < letterPool[letter]; runner++) {
+      letterArray.push(letter);
+    }
+  }
+  return letterArray;
+};
+
+const todaysSoup = letterSoup(letterPool);
+
 export const drawLetters = () => {
-  // Implement this method for wave 1
+  // array to store ten drawn letters
+  const tenLetters = [];
+
+  // loop to randomly populate tenLetters
+  while (tenLetters.length < 10) {
+    // choose randomLetter from randomized index from todaysSoup array, between 0 and todaysSoup length
+    let randomLetter =
+      todaysSoup[Math.floor(Math.random() * todaysSoup.length + 1)];
+
+    // check if randomLetter count within tenLetters is less than allotted letter quantity
+    // if true, add randomLetter to tenLetters array
+    if (collect(tenLetters).count(randomLetter) < letterPool[randomLetter]) {
+      tenLetters.push(randomLetter);
+    }
+  }
+
+  return tenLetters;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  const availableLetters = lettersInHand;
+  const upperInput = input.toUpperCase();
+
+  for (let i = 0; i < upperInput.length; i++) {
+    if (availableLetters.includes(upperInput[i])) {
+      availableLetters.shift(upperInput[i]);
+    } else {
+      return false;
+    }
+  }
+  return true;
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  let score = 0;
+  let upperWord = word.toUpperCase();
+
+  for (let letter of upperWord) {
+    if (letter in scoreChart) {
+      score += scoreChart[letter];
+    }
+  }
+
+  if (word.length >= 7) {
+    score += 8;
+  }
+  return score;
 };
 
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  let wordScores = {};
+  let highestScoring = [];
+
+  for (let word of words) {
+    wordScores[word] = scoreWord(word);
+  }
+
+  let maxScore = Math.max(...Object.values(wordScores));
+
+  console.log("word scores obj:", wordScores);
+  console.log("max score:", maxScore);
+
+  for (let word in wordScores) {
+    if (wordScores[word] === maxScore) {
+      highestScoring.push(word);
+    }
+  }
+
+  // sort highest scoring words by length, with shortest word at the beginning
+  highestScoring.sort((a, b) => a.length - b.length);
+  console.log(highestScoring);
+
+  for (let word of highestScoring) {
+    if (word.length === 10) {
+      return { word: word, score: maxScore };
+    }
```

Comment:
This works but you could simplify this by creating one variable to hold the high scoring data. Using two variables takes up more memory and most primitives (booleans, numbers, null, undefined, etc.) are constant (O(1)) space. This is because, no matter if true or false, or 1 or 1,000, the amount of space consumed stays constant. Reference types (arrays and objects) — Reference types require O(n) space as well. ’n’ being the length for an array, or in the case of an object, ’n’ being the amount of keys. Reference types are O(n) because the more elements in an array, or the more keys in an object, the more space they consume.

The `highestScoring` array does have a space complexity of 0(n), while the `wordScores` number has a space complexity of 0(1), but since that's less than the `highestScoring` variable, we drop it. You are also using a `maxScore` variable which holds the result of scoreWord call and you are creating this variable twice, which triggers the scoreWord function call twice.